File: src/schemas/metadata/MetadataSet.ts
import { ClassProperties } from "../../typings/ClassProperties";

import { IMetadataSet } from "./IMetadataSet";
import { IMetadataTypeTag } from "./IMetadataTypeTag";
import { Metadata } from "./Metadata";

export class MetadataSet {
  public readonly value: Metadata;
  public readonly tags: IMetadataTypeTag[][];
  private name_?: string;

  private constructor(props: ClassProperties<MetadataSet>) {
    this.value = props.value;
    this.tags = props.tags;
  }

  public static create(props: ClassProperties<MetadataSet>): MetadataSet {
    return new MetadataSet(props);
  }

  public getName(): string {
    return (this.name_ ??= (() => {
      const symbol: string = `Set<${this.value.getName()}>`;
      if (this.tags.length === 0) return symbol;
      else if (this.tags.length === 1) {
        const str: string = [symbol, ...this.tags[0]!.map((t) => t.name)].join(
          " & ",
        );
        return `(${str})`;
      }
      const rows: string[] = this.tags.map((row) => {
        const str: string = row.map((t) => t.name).join(" & ");
        return row.length === 1 ? str : `(${str})`;
      });
      return `(${symbol} & (${rows.join(" | ")}))`;
    })());
  }

  public toJSON(): IMetadataSet {
    return {
      value: this.value.toJSON(),
      tags: this.tags,
    };
  }
}

File: packages/typia/native/core/schemas/metadata/MetadataSet.go
package metadata

type IMetadataSchema_ISet struct {
  Value *IMetadataSchema
  Tags  [][]IMetadataTypeTag
}

type MetadataSet struct {
  Value         *MetadataSchema
  Tags          [][]IMetadataTypeTag
  name_         string
  display_name_ string
}

func MetadataSet_create(props MetadataSet) *MetadataSet {
  return &MetadataSet{
    Value: props.Value,
    Tags:  cloneTagMatrix(props.Tags),
  }
}

func (obj *MetadataSet) GetName() string {
  if obj.name_ == "" {
    obj.name_ = taggedName("Set<"+safeMetadataName(obj.Value)+">", obj.Tags)
  }
  return obj.name_
}

func (obj *MetadataSet) GetDisplayName() string {
  if obj.display_name_ == "" {
    obj.display_name_ = taggedName("Set<"+safeMetadataDisplayName(obj.Value)+">", obj.Tags)
  }
  return obj.display_name_
}

func (obj *MetadataSet) ToJSON() IMetadataSchema_ISet {
  return IMetadataSchema_ISet{
    Value: obj.Value.ToJSON(),
    Tags:  cloneTagMatrix(obj.Tags),
  }
}
